### ISSUE TITLE
preprocessing data with dictionary of "transformer_lm.wmt19.en": too many tokens are replaced by <unk>

### ISSUE THREAD

**Issue by Sohyo** [pytorch/fairseq#2299]
Hello

I'm trying to pre-process data using the given dictionary of wmt19en.
While doing this I noticed that wmt19en dictionary has **42022 types** and because of this, many of tokens are replaced by <unk>. Then this resulting in very high perplexity. How I can prevent replacing too many tokens to <unk>?

For instance, when I pre-process without using the given dictionary from wmt19en or using a dictionary from adaptive_lm_wiki103.v2, there are **267744 types**.  But then I can't use this because the amount of dictionary is different from transformer_lm.wmt19.en.

To see easy, I just try to pre-proecess wiki-103 dataset with the dictionary from wmt19.en, below ;

> `fairseq-preprocess     --only-source     --trainpref examples/language_model/wikitext-103/wiki.train.tokens     --validpref examples/language_model/wikitext-103/wiki.valid.tokens     --testpref examples/language_model/wikitext-103/wiki.test.tokens     --destdir data-bin/wikitext-for-wmt19 --srcdict data-bin/wmt19.en/dict.txt   --workers 20
`

Then, more than 17% of tokens are replaced to <unk>

> `2020-07-03 16:07:53 | INFO | fairseq_cli.preprocess | Namespace(align_suffix=None, alignfile=None, all_gather_list_size=16384, bf16=False, bpe=None, checkpoint_suffix='', cpu=False, criterion='cross_entropy', dataset_impl='mmap', destdir='data-bin/wikitext-for-wmt19', empty_cache_freq=0, fp16=False, fp16_init_scale=128, fp16_no_flatten_grads=False, fp16_scale_tolerance=0.0, fp16_scale_window=None, joined_dictionary=False, log_format=None, log_interval=100, lr_scheduler='fixed', memory_efficient_bf16=False, memory_efficient_fp16=False, min_loss_scale=0.0001, model_parallel_size=1, no_progress_bar=False, nwordssrc=-1, nwordstgt=-1, only_source=True, optimizer='nag', padding_factor=8, profile=False, quantization_config_path=None, seed=1, source_lang=None, srcdict='data-bin/wmt19.en/dict.txt', target_lang=None, task='translation', tensorboard_logdir='', testpref='examples/language_model/wikitext-103/wiki.test.tokens', tgtdict=None, threshold_loss_scale=None, thresholdsrc=0, thresholdtgt=0, tokenizer=None, tpu=False, trainpref='examples/language_model/wikitext-103/wiki.train.tokens', user_dir=None, validpref='examples/language_model/wikitext-103/wiki.valid.tokens', workers=20)
2020-07-03 16:07:53 | INFO | fairseq_cli.preprocess | [None] Dictionary: 42022 types
2020-07-03 16:09:04 | INFO | fairseq_cli.preprocess | [None] examples/language_model/wikitext-103/wiki.train.tokens: 1801350 sents, 103227021 tokens, 17.6% replaced by <unk>
2020-07-03 16:09:04 | INFO | fairseq_cli.preprocess | [None] Dictionary: 42022 types
2020-07-03 16:09:05 | INFO | fairseq_cli.preprocess | [None] examples/language_model/wikitext-103/wiki.valid.tokens: 3760 sents, 217646 tokens, 17.2% replaced by <unk>
2020-07-03 16:09:05 | INFO | fairseq_cli.preprocess | [None] Dictionary: 42022 types
2020-07-03 16:09:06 | INFO | fairseq_cli.preprocess | [None] examples/language_model/wikitext-103/wiki.test.tokens: 4358 sents, 245569 tokens, 17.6% replaced by <unk>
2020-07-03 16:09:06 | INFO | fairseq_cli.preprocess | Wrote preprocessed data to data-bin/wikitext-for-wmt19`

To sum up, 

1. Is there any way to prevent replacing too many tokens to <unk> ?

2. Also, is there any way to not add <unk> for using 'fairseq-eval-lm' ?

Thank you in advance!

**Comment by myleott** [comment]
That model has a byte-pair encoded (BPE) vocabulary, so you need to first apply BPE and then tokenize.

For example:

```python
en_lm = torch.hub.load('pytorch/fairseq', 'transformer_lm.wmt19.en')

# if you try to encode raw text you'll get unknowns
>>> en_lm.tgt_dict.string(en_lm.binarize("Hello world"))
'<unk> world'

# but if you first apply BPE
>>> en_lm.bpe.encode("Hello world")
'H@@ ello world'

# then it will work
>>> en_lm.tgt_dict.string(en_lm.binarize("H@@ ello world"))
'H@@ ello world'

# you can remove the BPE by replacing "@@ " with "" or calling bpe.decode
>>> en_lm.bpe.decode("H@@ ello world")
'Hello world'
```
